### PR TITLE
[OS Load Balancers] Allow Service Annotations to override the settings specified in --cloud-config

### PR DESCRIPTION
**Pull request by hacktastic**
tl;dr: Added a set of Service Annotations that allow the Service author to directly control key settings of the load balancers he or she creates

This is my proposal for issue #40054 

There is a definite lack of flexibility when it comes to configuring OpenStack load balancers through k8s Services. The following settings may be set in an administrator's --cloud-config file (under the [LoadBalancer] heading), but they apply to all Loadbalancer-type Services on your cluster:

lb-version
subnet-id
floating-network-id
lb-method
create-monitor
monitor-delay
monitor-timeout
monitor-max-retries
manage-security-groups
node-security-group

Furthermore, there is no setting for LB "provider". I have added the "provider" setting to the list of available --cloud-config settings, as well as added some logic to the openstack loadbalancer driver that will allow the user to control almost all of these settings via Service Annotations.

The annotations all share a common, simple structure that is similar to that of pkg/cloudprovider/providers/aws's annotations:
service.beta.kubernetes.io/openstack-load-balancer-provider
service.beta.kubernetes.io/openstack-load-balancer-subnet-id
service.beta.kubernetes.io/openstack-load-balancer-floating-network-id
service.beta.kubernetes.io/openstack-load-balancer-lb-method
service.beta.kubernetes.io/openstack-load-balancer-create-monitor
service.beta.kubernetes.io/openstack-load-balancer-monitor-delay
service.beta.kubernetes.io/openstack-load-balancer-monitor-timeout
service.beta.kubernetes.io/openstack-load-balancer-monitor-max-retries
service.beta.kubernetes.io/openstack-load-balancer-manage-security-groups
service.beta.kubernetes.io/openstack-load-balancer-node-security-group

Where appropriate, I have inserted code into pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go to make sure that the Annotations take a higher priority than their corresponding --cloud-config settings.

### CODE REVIEW

**Comment by hacktastic** [comment]
I signed it!

**Comment by liggitt** [comment]
assume a cluster-admin doesn't want their users to be able to override the settings in cloud config. how would a cluster-admin control what API users are allowed to do?

**Comment by hacktastic** [comment]
@liggitt, good point! How about something like this?

`type LoadBalancerOpts struct {
	LBVersion            string     `gcfg:"lb-version"` // overrides autodetection. v1 or v2
	SubnetId             string     `gcfg:"subnet-id"`  // required
	FloatingNetworkId    string     `gcfg:"floating-network-id"`
	LBMethod             string     `gcfg:"lb-method"`
	CreateMonitor        bool       `gcfg:"create-monitor"`
	MonitorDelay         MyDuration `gcfg:"monitor-delay"`
	MonitorTimeout       MyDuration `gcfg:"monitor-timeout"`
	MonitorMaxRetries    uint       `gcfg:"monitor-max-retries"`
	ManageSecurityGroups bool       `gcfg:"manage-security-groups"`
	NodeSecurityGroupID  string     `gcfg:"node-security-group"`
	Provider             string     `gcfg:"provider"`

	// Optional settings that allow a cluster admin to prevent his users from setting LB parameters through Service Annotations
	DisableAnnotationLBVersion            bool `gcfg:"disable-annotation-lb-version"`
	DisableAnnotationSubnetId             bool `gcfg:"disable-annotation-subnet-id"`
	DisableAnnotationFloatingNetworkId    bool `gcfg:"disable-annotation-floating-network-id"`
	DisableAnnotationLBMethod             bool `gcfg:"disable-annotation-lb-method"`
	DisableAnnotationCreateMonitor        bool `gcfg:"disable-annotation-create-monitor"`
	DisableAnnotationMonitorDelay         bool `gcfg:"disable-annotation-monitor-delay"`
	DisableAnnotationMonitorTimeout       bool `gcfg:"disable-annotation-monitor-timeout"`
	DisableAnnotationMonitorMaxRetries    bool `gcfg:"disable-annotation-monitor-max-retries"`
	DisableAnnotationManageSecurityGroups bool `gcfg:"disable-annotation-manage-security-groups"`
	DisableAnnotationNodeSecurityGroupID  bool `gcfg:"disable-annotation-node-security-group"`
	DisableAnnotationProvider             bool `gcfg:"disable-annotation-provider"`
}`

**Comment by liggitt** [comment]
not sure, I'd let those better acquainted with cloudconfig comment on it. a few general thoughts on the security aspect of exposing this in the API:
* default to secure... make the cluster-admin opt in to exposing knobs to end users
* consider whether you'd want this power to be available in some namespaces, but not others

**Comment by hacktastic** [comment]
@anguslees any thoughts on this PR?

**Comment by xsgordon** [comment]
@liggitt can you take a look at the latest patches here?

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 1]
```diff
@@ -84,6 +84,20 @@ type LoadBalancerOpts struct {
 	MonitorMaxRetries    uint       `gcfg:"monitor-max-retries"`
 	ManageSecurityGroups bool       `gcfg:"manage-security-groups"`
 	NodeSecurityGroupID  string     `gcfg:"node-security-group"`
+	Provider             string     `gcfg:"provider"`
+
+	// Optional settings that allow a cluster admin to allow his users to set specific LB parameters through Service Annotations
+	EnableAnnotationLBVersion            bool `gcfg:"disable-annotation-lb-version"`
```
Comment (liggitt) [inline pkg/cloudprovider/providers/openstack/openstack.go]: annotation is backwards from the field name

Comment (hacktastic) [inline pkg/cloudprovider/providers/openstack/openstack.go]: Oops! Good spot. I've pushed a new commit to make the field tags line up with their respective vars.

**Comment by hacktastic** [comment]
What else does this PR need to get merged?

**Comment by xsgordon** [comment]
@k8s-bot test this

**Comment by liggitt** [comment]
@kubernetes/api-reviewers opinion on this as a method for configuring provider-specific knobs on services?

**Comment by smarterclayton** [comment]
Purely stylistic: "openstack.service.beta.../load-balancer-..." or "openstack.lb.service.beta.../max-retries" might be better subdivision.  Are these truly beta, or more alpha?

If this is net new, I'm willing to give this a try.  We cannot assume forever that LB are going to be limited, and if this is alpha and helps us gather useful feedback about this as a general approach for extension.  The flags to control the parameters are important - although the more options we add the more flags will be required.

Has someone involved in the ingress annotations discussions had a chance to weigh in?  If not it would be good to get some consensus.

Are there a larger number of parameters after this that are likely to get added?

**Comment by thockin** [comment]
stylistically, if these are openstack related the annotation should probably be of the form `loadbalancer.openstack.org/foobar` - there's no need for them to be in the kubernetes.io domain.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 2]
```diff
@@ -49,6 +50,18 @@ import (
 const loadbalancerActiveTimeoutSeconds = 120
 const loadbalancerDeleteTimeoutSeconds = 30
 
+// svcAnnotationProvider and svcAnnotationSubnetID are service annotations that override their corresponding default (i.e. from cloud-config) settings
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: This block will now become the canonical place where these options are documented - please add comments to each describing expected values, behaviour that the option affects, etc.  (Only has to be a line or two, see the AWS cloud provider for examples)

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 3]
```diff
@@ -532,18 +545,81 @@ func createNodeSecurityGroup(client *gophercloud.ServiceClient, nodeSecurityGrou
...
 }
 
+//getStringFromServiceAnnotation searches a given v1.Service for a specific annotationKey and either returns the annotation's value or a specified defaultSetting
+func getStringFromServiceAnnotation(service *v1.Service, annotationKey string, defaultSetting string, enabledFlag bool) string {
+	glog.V(4).Infof("getStringFromServiceAnnotation(%v, %v, %v, %v)", service, annotationKey, defaultSetting, enabledFlag)
+	if !enabledFlag {
+		return defaultSetting
+	}
+	if annotationValue, ok := service.Annotations[annotationKey]; ok {
+		//if there is an annotation for this setting, set the "setting" var to it
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: This comment seems incorrect - there is no `setting` var

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 4]
```diff
@@ -532,18 +545,81 @@ func createNodeSecurityGroup(client *gophercloud.ServiceClient, nodeSecurityGrou
...
+		glog.V(4).Infof("Found a Service Annotation: %v = %v", annotationKey, annotationValue)
+		return annotationValue
+	}
+	//if there is no annotation, set "settings" var to the value from cloud config
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: ditto - no `settings` var

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 5]
```diff
@@ -532,18 +545,81 @@ func createNodeSecurityGroup(client *gophercloud.ServiceClient, nodeSecurityGrou
...
+	glog.V(4).Infof("Could not find a Service Annotation; falling back on cloud-config setting: %v = %v", annotationKey, defaultSetting)
+	return defaultSetting
+}
+
+//getBoolFromServiceAnnotation searches a given v1.Service for a specific annotationKey and either returns the annotation's value or a specified defaultSetting
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: There's a reasonable amount of duplicated code here and in the Duration version below.  I think I'd rather we just called `getStringFromServiceAnnotation` to do the heavy lifting in each of these, and then parsed the (string) result.  I don't care whether you take the default value in string form always, or take a properly typed value and then convert it to a string (somewhat unnecessarily) for `getStringFromServiceAnnotation`.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 6]
```diff
@@ -754,14 +836,31 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
...
+		createMonitor := getBoolFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerCreateMonitor, lbaas.opts.CreateMonitor, lbaas.opts.EnableAnnotationCreateMonitor)
+
+		// if this service has an annotation for MonitorDelay, use that instead
+		monitorDelay := getDurationFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerMonitorDelay, lbaas.opts.MonitorDelay, lbaas.opts.EnableAnnotationMonitorDelay)
+
+		// if this service has an annotation for MonitorTimeout, use that instead
+		monitorTimeout := getDurationFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerMonitorTimeout, lbaas.opts.MonitorTimeout, lbaas.opts.EnableAnnotationMonitorTimeout)
+
+		// if this service has an annotation for MonitorMaxRetries, use that instead
+		_monitorMaxRetries := getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerMonitorMaxRetries, strconv.FormatUint(uint64(lbaas.opts.MonitorMaxRetries), 10), lbaas.opts.EnableAnnotationMonitorMaxRetries)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Looks like we need a `getUint64FromServiceAnnotation` too (also used below)

**Review by anguslees — CHANGES_REQUESTED** [review]
1. Thanks, this is a needed improvement
2. I hate all the conditional "enable" stuff ;)  Imo we should just follow the AWS precedent that everyone seems ok with: move everything to service annotations and let the user break their setup if they use the wrong values.

I think the only value that makes sense to set cluster-wide is `lb-version` and maybe `node-security-group`.  And the only reason we have config options for these two is because we're lazy - we should be able to reliably automatically determine both.

**Comment by liggitt** [comment]
> Imo we should just follow the AWS precedent that everyone seems ok with: move everything to service annotations and let the user break their setup if they use the wrong values.

Don't assume the cluster admin and the service creator/annotator are the same user. Anything that these annotations expose that could cost/disrupt/control the load balancer that wasn't previously possible should be opted into by the cluster admin.

**Comment by anguslees** [comment]
@liggitt: I agree with the sentiment, but I disagree that we should pioneer this approach with the openstack cloud provider.  I think we should be following the same sorts of approach that the other cloud provider plugins are doing, and they (notably aws and gce) allow the service creator/annotator to heavily influence the LB at the service annotation level (including triggering $$ costs and breaking the setup).  Right now, the openstack provider is the "odd one out" in terms of where this sort of configuration exists (for historical and ease-of-implementation reasons).

If we want to enforce a separation of the roles as you're describing, then lets do that(!) - but in a coordinated way across all cloud providers.

**Comment by liggitt** [comment]
If we are ever going to make headway on letting the cluster admin have control in such an environment, we can't keep adding power via uncontrolled channels like annotations and just say "we'll lock it down later"

This should have been considered before the other providers did similar things, and the fact that it wasn't doesn't seem like a good reason not to consider this now. 

That said, a lighter weight way of toggling which features are honored would appeal to me, so a cluster admin could easily indicate they wanted *all* annotation-driven features enabled on their services with an `enableAnnotationOverrides: "*"`, or pick a subset with `enableAnnotationOverrides: "lb-version,lb-method,..."`, etc

**Comment by liggitt** [comment]
Put another way, with an openstack provider in 1.6, a cluster admin today can give users edit access to services without them having this control. We should not default to escalating those users in 1.7 without the admin taking positive action to indicate that is desired.

**Comment by anguslees** [comment]
@liggitt: ok, so here's the current list:

- provider (added new in this PR)
- lb-version
- subnet-id
- floating-network-id
- lb-method
- create-monitor
- monitor-delay
- monitor-timeout
- monitor-max-retries
- manage-security-groups
- node-security-group

I suggest we (as a group), go through each of these and work out:
- Which ones can/should be removed entirely since they can be automatically determined
- Which ones need to vary by service and should be moved to annotations
- Which ones are cluster-wide or the breakage is not isolated to a single service and so should be kept (only) in provider config

Here's my stab at going through that process for each of the above, but please followup and I'll update this comment with any corrections:

### Automate away entirely:
- lb-version (keep the existing "auto" behaviour and just remove the manual overrides)
- node-security-groups (find security groups on all k8s nodes)

### Per-service annotation:
- provider
- subnet-id  (*)
- floating-network-id  (*)
- lb-method
- create-monitor
- monitor-delay
- monitor-timeout
- monitor-max-retries

### Cluster-wide config:
- manage-security-groups (or just always do it?)

## The bits I'm not happy about:

- (*)  These need to vary per-service so must be moved to annotations, but they could lead to "odd" results if the wrong value is given.  Is there any way we can automate these away or reduce the selection to only "safe" choices?

- The "monitor" options seem a bit superfluous, can we just declare some values and always create the monitor?

**Comment by dims** [comment]
@k8s-bot ok to test

**Comment by liggitt** [comment]
> I suggest we (as a group), go through each of these and work out:
> 
>     Which ones can/should be removed entirely since they can be automatically determined
>     Which ones need to vary by service and should be moved to annotations
>     Which ones are cluster-wide or the potential breakage is not isolated to a single service and so should be kept (only) in provider config

Those are great questions. I'm happy to defer to those with more domain knowledge about which bucket each setting falls into.

For the settings exposed per-service via annotations, I'd ask additional questions (and would want the implications to be part of the documentation for enabling the settings):
* could this be used to negatively impact other services or other namespaces?
* could this be used to redirect traffic to/access traffic from otherwise inaccessible networks/addresses? (e.g. end-run around network policy)
* could this increase consumption of a scarce and/or expensive resource?

For all those reasons (and since service definitions annotated with provider-specific settings are not portable between providers), I think it makes sense to let admins decide whether to enable those settings.

**Comment by xsgordon** [comment]
In the hope of trying to garner some feedback I copied @anguslees' proposal to https://etherpad.openstack.org/p/service-annotations-for-kubernetes-cloud-config and sent a note to the sig-openstack list: https://groups.google.com/forum/#!topic/kubernetes-sig-openstack/6GBeADqM-g8

**Comment by xsgordon** [comment]
keep-open

**Comment by zetaab** [comment]
Oh there is already PR what I need. I created new one https://github.com/kubernetes/kubernetes/pull/49697 but this looks very promissing. Could we get progress here? @hacktastic at least I need possibility to configure floating-ids in annotations.

**Comment by FengyunPan** [comment]
@hacktastic Please rebase this PR.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 7]
```diff
@@ -501,18 +514,81 @@ func createNodeSecurityGroup(client *gophercloud.ServiceClient, nodeSecurityGrou
 	return nil
 }
 
+//getStringFromServiceAnnotation searches a given v1.Service for a specific annotationKey and either returns the annotation's value or a specified defaultSetting
+func getStringFromServiceAnnotation(service *v1.Service, annotationKey string, defaultSetting string, enabledFlag bool) string {
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Repeat with Line 559

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 8]
```diff
@@ -746,7 +828,7 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
 				_, err := v2pools.CreateMember(lbaas.network, pool.ID, v2pools.CreateMemberOpts{
 					ProtocolPort: int(port.NodePort),
 					Address:      addr,
-					SubnetID:     lbaas.opts.SubnetId,
+					SubnetID:     getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerSubnetID, lbaas.opts.SubnetId, lbaas.opts.EnableAnnotationSubnetId),
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: We have supported for autoprobing subnet-id, please do a rebase for it, get more info from Line 609.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 9]
```diff
@@ -1116,8 +1221,16 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *v1.
 		return nil
 	}
 
-	if loadbalancer != nil && loadbalancer.VipPortID != "" {
+	// if this service has an annotation for floating network ID, use that instead of the default
+	floatingNetworkID := getStringFromServiceAnnotation(service, ServiceAnnotationLoadBalancerFloatingNetworkID, lbaas.opts.FloatingNetworkId, lbaas.opts.EnableAnnotationFloatingNetworkId)
+
+	if floatingNetworkID != "" && loadbalancer != nil && loadbalancer.VipPortID != "" {
 		portID := loadbalancer.VipPortID
+		port, err := getPortByIP(lbaas.network, loadbalancer.VipAddress)
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Bad code, please remove it, we can get portID by loadbalancer.VipPortID not getPortByIp().

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 10]
```diff
@@ -1116,8 +1221,16 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *v1.
 		return nil
 	}
 
-	if loadbalancer != nil && loadbalancer.VipPortID != "" {
+	// if this service has an annotation for floating network ID, use that instead of the default
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Using "if loadbalancer != nil && loadbalancer.VipPortID != """ is enough

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 11]
```diff
@@ -1386,7 +1524,7 @@ func (lb *LbaasV1) EnsureLoadBalancer(clusterName string, apiService *v1.Service
 		Protocol:     "TCP",
 		ProtocolPort: int(ports[0].Port), //TODO: need to handle multi-port
 		PoolID:       pool.ID,
-		SubnetID:     lb.opts.SubnetId,
+		SubnetID:     subnetID,
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Ditto.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 12]
```diff
@@ -772,14 +854,31 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
...
+		createMonitor := getBoolFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerCreateMonitor, lbaas.opts.CreateMonitor, lbaas.opts.EnableAnnotationCreateMonitor)
+
+		// if this service has an annotation for MonitorDelay, use that instead
+		monitorDelay := getDurationFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerMonitorDelay, lbaas.opts.MonitorDelay, lbaas.opts.EnableAnnotationMonitorDelay)
+
+		// if this service has an annotation for MonitorTimeout, use that instead
+		monitorTimeout := getDurationFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerMonitorTimeout, lbaas.opts.MonitorTimeout, lbaas.opts.EnableAnnotationMonitorTimeout)
+
+		// if this service has an annotation for MonitorMaxRetries, use that instead
+		_monitorMaxRetries := getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerMonitorMaxRetries, strconv.FormatUint(uint64(lbaas.opts.MonitorMaxRetries), 10), lbaas.opts.EnableAnnotationMonitorMaxRetries)
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: MonitorMaxRetries is not a string.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 6 above)
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: +1

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 13]
```diff
@@ -70,6 +71,18 @@ const (
...
+const ServiceAnnotationLoadBalancerSubnetID = "service.beta.kubernetes.io/openstack-load-balancer-subnet-id"
+const ServiceAnnotationLoadBalancerFloatingNetworkID = "service.beta.kubernetes.io/openstack-load-balancer-floating-network-id"
+const ServiceAnnotationLoadBalancerLBMethod = "service.beta.kubernetes.io/openstack-load-balancer-lb-method"
+const ServiceAnnotationLoadBalancerCreateMonitor = "service.beta.kubernetes.io/openstack-load-balancer-create-monitor"
+const ServiceAnnotationLoadBalancerMonitorDelay = "service.beta.kubernetes.io/openstack-load-balancer-monitor-delay"
+const ServiceAnnotationLoadBalancerMonitorTimeout = "service.beta.kubernetes.io/openstack-load-balancer-monitor-timeout"
+const ServiceAnnotationLoadBalancerMonitorMaxRetries = "service.beta.kubernetes.io/openstack-load-balancer-monitor-max-retries"
+const ServiceAnnotationLoadBalancerManageSecurityGroups = "service.beta.kubernetes.io/openstack-load-balancer-manage-security-groups"
+const ServiceAnnotationLoadBalancerNodeSecurityGroupID = "service.beta.kubernetes.io/openstack-load-balancer-node-security-group-id"
```
Comment (zhouhaibing089) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: const list should be something like:

```
const (
    Name = "value"
    ...
)
```

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 14]
```diff
@@ -1329,16 +1442,19 @@ func (lb *LbaasV1) EnsureLoadBalancer(clusterName string, apiService *v1.Service
...
-	if lbmethod == "" {
-		lbmethod = pools.LBMethodRoundRobin
+	// if this service has an annotation for LB method, use that instead
+	lbMethod := pools.LBMethod(getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerLBMethod, lb.opts.LBMethod, lb.opts.EnableAnnotationLBMethod))
+
+	if lbMethod == "" {
+		lbMethod = pools.LBMethodRoundRobin
 	}
 	name := cloudprovider.GetLoadBalancerName(apiService)
+	subnetID := getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerSubnetID, lb.opts.SubnetId, lb.opts.EnableAnnotationSubnetId)
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: currently there are three ways to get subnet-id:
1.lbaas.opts.SubnetId
2.ServiceAnnotationLoadBalancerSubnetID
3.autoprobe subnet-id by "getSubnetIDForLB()"
You should also consider 3. If 1 and 2 are empty, we can use 3.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 15]
```diff
@@ -1080,7 +1186,7 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *v1.Service
 			_, err := v2pools.CreateMember(lbaas.network, pool.ID, v2pools.CreateMemberOpts{
 				Address:      addr,
 				ProtocolPort: int(port.NodePort),
-				SubnetID:     lbaas.opts.SubnetId,
+				SubnetID:     getStringFromServiceAnnotation(service, ServiceAnnotationLoadBalancerSubnetID, lbaas.opts.SubnetId, lbaas.opts.EnableAnnotationSubnetId),
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Currently there are three ways to get subnet-id:
1.lbaas.opts.SubnetId
2.ServiceAnnotationLoadBalancerSubnetID
3.autoprobe subnet-id by "getSubnetIDForLB()"
You should also consider 3. If 1 and 2 are empty, we can use 3.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 3 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: +1

**Review by FengyunPan — COMMENTED** [review]
On my side, Please do a change again, thank you:
1. address our comments
2. do a rebase
3. I see many commits in your PR, but some commit is useless. Can you combine them together? Or let them be simple?

**Comment by FengyunPan** [comment]
@hacktastic I love this PR, do you have spare time to address our comment? If yes, I will add lgtm label to this PR.

**Comment by dims** [comment]
/close

please let me know if you wish to rebase/reopen this PR.
